mobile_back_apple/cpp/backend_coreml/main.cc

```diff
@@ -0,0 +1,129 @@
+#import <CoreML/CoreML.h>
+
+#include <cstring>
+
+#include "coreml_settings.h"
+#include "coreml_util.h"
+#include "flutter/cpp/c/backend_c.h"
+#include "flutter/cpp/c/type.h"
+
+struct CoreMLBackendData {
+  const char *name = "Core ML";
+  const char *vendor = "Apple";
+  const char *accelerator = "Neural Engine";
+  MLModel *model{nullptr};
+  uint32_t real_batch_size = 1;
+  int32_t original_tensor_size = 0;
+};
+
+static bool backendExists = false;
+
+// Return the name of the backend
+const char *mlperf_backend_vendor_name(mlperf_backend_ptr_t backend_ptr) {
+  return ((CoreMLBackendData *)backend_ptr)->vendor;
+}
+
+// Return the name of the accelerator.
+const char *mlperf_backend_accelerator_name(mlperf_backend_ptr_t backend_ptr) {
+  return ((CoreMLBackendData *)backend_ptr)->accelerator;
+}
+
+// Return the name of this backend.
+const char *mlperf_backend_name(mlperf_backend_ptr_t backend_ptr) {
+  return ((CoreMLBackendData *)backend_ptr)->name;
+}
+
+// Should return true if current hardware is supported.
+bool mlperf_backend_matches_hardware(const char **not_allowed_message,
+                                     const char **settings,
+                                     const mlperf_device_info_t *device_info) {
+  (void)device_info;
+  *settings = coreml_settings.c_str();
+  *not_allowed_message = "This is a stub backend for Core ML";
+  return true;
+}
+
+// Create a new backend and return the pointer to it.
+mlperf_backend_ptr_t mlperf_backend_create(
+    const char *model_path, mlperf_backend_configuration_t *configs,
+    const char *native_lib_path) {
+
+  // Test calling Objective-C method
+  CoreMLDelegate *coremlDelegate = [[CoreMLDelegate alloc] init];
+  [coremlDelegate hello];
```

Comment:
@freedomtan Do you know how can I call Objective-C methods in C++?

I got this error when running:
```
2022-07-08 09:43:31.798896: I flutter/cpp/binary/main.cc:129] Using External backend
2022-07-08 09:43:31.834418: E flutter/cpp/backends/external.cc:140] Unable to load: bazel-bin/mobile_back_apple/cpp/backend_coreml/libcoremlbackend.so: dlopen(bazel-bin/mobile_back_apple/cpp/backend_coreml/libcoremlbackend.so, 0x0005): symbol not found in flat namespace '_OBJC_CLASS_$_CoreMLDelegate'
```

Reply:
as you can guess, `symbol not found in flat namespace '_OBJC_CLASS_$_CoreMLDelegate` means it is not in `bazel-bin/mobile_back_apple/cpp/backend_coreml/libcoremlbackend.so`. Please check if that's true (with tools such as `nm, objdump, or otool`) first.

Reply:
I have this:
```shell
➜ nm bazel-bin/mobile_back_apple/cpp/backend_coreml/libcoremlbackend.so  
                 U _NSLog
                 U _OBJC_CLASS_$_CoreMLDelegate
                 U _OBJC_CLASS_$_MLModel
                 U _OBJC_CLASS_$_NSString
                 U _OBJC_CLASS_$_NSURL
                 U __ZNSt3__112basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEED1Ev
                 U __Znwm
                 U ___CFConstantStringClassReference
                 U ___cxa_atexit
                 U _memcpy
00000000000035c8 T _mlperf_backend_accelerator_name
0000000000003608 T _mlperf_backend_create
000000000000371c T _mlperf_backend_delete
0000000000003728 T _mlperf_backend_flush_queries
0000000000003730 T _mlperf_backend_get_input_count
00000000000037bc T _mlperf_backend_get_input_type
0000000000003868 T _mlperf_backend_get_output
00000000000037d0 T _mlperf_backend_get_output_count
000000000000385c T _mlperf_backend_get_output_type
0000000000003720 T _mlperf_backend_issue_query
00000000000035d8 T _mlperf_backend_matches_hardware
00000000000035d0 T _mlperf_backend_name
00000000000037c8 T _mlperf_backend_set_input
00000000000035c0 T _mlperf_backend_vendor_name
                 U _objc_alloc_init
                 U _objc_msgSend
```

Reply:
FYI, I have limited experience with C/C++ or Objective-C so I don't know how to debug / interpret this.

Reply:
U stands for undefined. You can find that in [`man nm`](https://www.unix.com/man-page/osx/1/nm/).
`U _OBJC_CLASS_$_CoreMLDelegate` means, hey, I know you want to use `_OBJC_CLASS_$_CoreMLDelegate`, but I don't know where it is.

Let's use C as an example. If we `main.c` and `foo.c` and we want  foo.c's void foo() in `main.c`. We declare `extern foo()` in `main.c` and call it. When we compile `main.c` to `main.o` (`cc -c main.c`) and examine `main.o` with `nm`, we'll see something like 'U foo'.

OK. In plain English, the code of `_OBJC_CLASS_$_CoreMLDelegate` is not linked into `bazel-bin/mobile_back_apple/cpp/backend_coreml/libcoremlbackend.so` (I don't know why. That's why nm told us :-))

Reply:
I guess the culprit is https://github.com/mlcommons/mobile_app_open/blob/master/flutter/cpp/c/exported_symbols.lds.

Reply:
It turns out. It's broken in `//mobile_back_apple/cpp/backend_coreml:coreml_c` already
```
bazel build -c opt --cxxopt=-std=c++14 --host_cxxopt=-std=c++14 --cxxopt=-xobjective-c++ //mobile_back_apple/cpp/backend_coreml:coreml_util
```
you'll see something like, 
```
...
INFO: Found 1 target...
Target //mobile_back_apple/cpp/backend_coreml:coreml_util up-to-date:
  bazel-out/applebin_macos-darwin_x86_64-opt-ST-cd2b3b8d4835/bin/mobile_back_apple/cpp/backend_coreml/libcoreml_util.a
INFO: Elapsed time: 0.133s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
```
When we check the `libcoreml_util.a`, we can see
```
...
0000000000000128 S _OBJC_CLASS_$_CoreMLDelegate
...
```

Then, we check `//mobile_back_apple/cpp/backend_coreml:coreml_c`
```
bazel build -c opt --cxxopt=-std=c++14 --host_cxxopt=-std=c++14 --cxxopt=-xobjective-c++ //mobile_back_apple/cpp/backend_coreml:coreml_c
```
If we check the output `bazel-bin/mobile_back_apple/cpp/backend_coreml/libcoreml_c.lo`,
```
                 U _OBJC_CLASS_$_CoreMLDelegate
```
which means, `_OBJC_CLASS_$_CoreMLDelegate` is gone already.

Reply:
if we move `coreml_util.mm` to `coreml_util.cc` and include it in `coreml_c`, you can see `_OBJC_CLASS_$_CoreMLDelegate` is there. And then, we have to change `https://github.com/mlcommons/mobile_app_open/blob/master/flutter/cpp/c/exported_symbols.lds` to export it.

Reply:
How about changing Clang setting `Compile Sources As` from `According to File Type`  to `Objective-C++`.

Reply:
Thanks, that works.